iam_cert: allow paths to be specified so slurp is not necessary for remote host…

s-hertel opened the pull request:
…s to use

##### SUMMARY
In response to https://github.com/ansible/ansible/pull/24206
Allow the options cert, key, and cert_chain to allow either a path or a lookup.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/iam_cert.py

##### ANSIBLE VERSION
```
2.4.0
```

s-hertel committed:
allow paths to be specified so slurp is not necessary for remote hosts to use

s-hertel commented:
cc @ProjectRice

ProjectRice commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -225,14 +249,28 @@ def cert_action(module, iam, name, cpath, new_name, new_path, state,
             module.exit_json(changed=changed, msg='Certificate with the name %s already absent' % name)
 
 
+def read_files(cert, cert_body, key, key_body, cert_chain, cert_chain_body):
```
Maybe we could call this something like load_data or load content?   since it may not actually be reading files at all.

s-hertel replied:
Thanks for reviewing! I think I made all the changes. How does this look now?

ProjectRice commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -225,14 +249,28 @@ def cert_action(module, iam, name, cpath, new_name, new_path, state,
         state=dict(required=True, choices=['present', 'absent']),
         name=dict(),
         cert=dict(),
+        cert_body=dict(),
```
I don't mind the idea of 2 params, but prefer the option that was discussed whereby the same param was used regardless.  If it was a path/file (os.path.isfile) then load the file, otherwise treat as string.
https://github.com/ansible/ansible/pull/20787
Not sure what is "normal" here currently though?

ProjectRice commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -225,14 +249,28 @@ def cert_action(module, iam, name, cpath, new_name, new_path, state,
 
+def read_files(cert, cert_body, key, key_body, cert_chain, cert_chain_body):
+    # if paths are provided rather than lookups read the files and return the contents
+    if not cert_body and os.path.exists(cert):
```
As previously discussed on https://github.com/ansible/ansible/pull/20787 I much prefer os.path.isfile
That way it won't try to read a directory path.

ProjectRice commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -225,14 +249,28 @@ def cert_action(module, iam, name, cpath, new_name, new_path, state,
         cert=dict(),
+        cert_body=dict(),
         key=dict(no_log=True),
+        key_body=dict(no_log=True),
```
As per cert_body comment.

ProjectRice commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -225,14 +249,28 @@ def cert_action(module, iam, name, cpath, new_name, new_path, state,
         key=dict(no_log=True),
+        key_body=dict(no_log=True),
         cert_chain=dict(),
+        cert_chain_body=dict(),
```
As per cert_body comment.

ProjectRice reviewed (changes requested):
This probably also needs to be rebased against https://github.com/ansible/ansible/pull/26087

s-hertel committed:
Make requested changes

remove unused parameters

ProjectRice commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -222,6 +231,17 @@ def cert_action(module, iam, name, cpath, new_name, new_path, state,
+        key = open(key, 'r').read().rstrip()
+    if os.path.isfile(cert_chain):
+        cert_chain = open(module.params.get('cert_chain'), 'r').read()
+    return cert, key, cert_chain
```
Much nicer solution :+1:

ProjectRice commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -50,19 +50,19 @@
-      - Note that prior to 2.4, this parameter expected a path to a file.
-        Since 2.4 this is now accomplished using a lookup plugin. See examples for detail.
+      - The path to the CA certificate chain in PEM encoded format.
+        As of 2.4 this can also be accomplished using a lookup plugin.
```
I don't think this needs to necessarily mention lookup anymore because we're not forcing people to use content and they can just pass the file if they have it.   I believe the comment was only added because the module went from supporting only "content" to supporting only paths.  

Perhaps we're better off saying that as of 2.4, both content and paths are supported;  And if the param is not a path it is assumed to be content.

s-hertel replied:
This comment was because the module went from supporting only paths to only supporting content.
I'm fine changing this if it doesn't sense, but maybe you can phrase that for me? I feel like this is already essentially saying that as of 2.4 content and paths are permitted. "... can also be accomplished using a lookup plugin."

ProjectRice replied:
I've used comments like this on the local copy I'm using.
Happy to change the formatting or add stuff like "see examples"

```
     - The path to, or content of the certificate body in PEM encoded format.
        As of 2.4 both path and content are accepted.  If the parameter is not a file, it is assumed to be content.
```

```
     - The path to, or content of the private key in PEM encoded format.
        As of 2.4 both path and content are accepted.  If the parameter is not a file, it is assumed to be content. 
```

```
     - The path to, or content of the CA certificate chain in PEM encoded format.
        As of 2.4 both path and content are accepted.  If the parameter is not a file, it is assumed to be content.
```

s-hertel replied:
Thanks! I'm modifying it slightly because it's not clear which to use for versions prior to 2.4

ProjectRice commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -222,6 +231,17 @@ def cert_action(module, iam, name, cpath, new_name, new_path, state,
+    if os.path.isfile(key):
+        key = open(key, 'r').read().rstrip()
+    if os.path.isfile(cert_chain):
+        cert_chain = open(module.params.get('cert_chain'), 'r').read()
```
Missed one..    Should just be as below unless you want to also pass module into the function.

```
if os.path.isfile(cert_chain):
    cert_chain = open(cert_chain, 'r').read()
```

ProjectRice replied:
Does this also need an rstrip?

ryansb replied:
`module` isn't valid in this scope, you need to use the `cert_chain` parameter.

ProjectRice replied:
Something else to consider.    These statements should probably also check the value is not None, to avoid issues with isfile.  TypeError: coercing to Unicode: need string or buffer, NoneType found

```
if cert_chain is not None and os.path.isfile(cert_chain):
```

ProjectRice commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -222,6 +231,17 @@ def cert_action(module, iam, name, cpath, new_name, new_path, state,
+        cert_chain = open(module.params.get('cert_chain'), 'r').read()
+    return cert, key, cert_chain
+
+
```
Could probably get rid of the surplus required=False entries I seem to have missed as well

ryansb commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -269,9 +289,9 @@ def main():
-        cert = module.params.get('cert')
-        key = module.params.get('key')
-        cert_chain = module.params.get('cert_chain')
+        cert, key, cert_chain = load_data(cert=module.params.get('cert'),
```
There is one downside to this approach: if a user passes a nonexistent file path, this module will error out with some variation of `400: invalid certificate`. In that case, we need to make sure the error message surfaces the information about what certs were invalid. 

A way to check this might be if the "cert" is less than 200 characters, we can be certain it's supposed to be a file path since the server certs have to be at least 1024 bit. Over that, it's a maybe, but we should print out the cert that's supposedly "invalid" so people can tell what's wrong.

ProjectRice replied:
Can we cater for this checking if cert etc is not None somewhere else?

ryansb replied:
No, because someone could fat-finger a path such as `secretsss/cert.pem` when really the directory is called `secrets`. Then the `isfile` would fail and the module would try to upload the *string of the broken path* `secretsss/cert.pem` as the actual cert body (which would fail)

ProjectRice replied:
And unfortunately it looks like iam doens't require you to have the PEM headers.  At least via the gui anyway, so assuming its the same for CLI

ProjectRice replied:
I stand corrected, uploading certs/keys in the AWS GUI without the PEM headers does not work.

Perhaps we could use their regex pattern as per the [API reference](https://docs.aws.amazon.com/IAM/latest/APIReference/API_UploadServerCertificate.html)
or  simply check for the headers?  eg -----BEGIN

ProjectRice replied:
I've also double checked what happens in the scenario you mentioned, and below is the result I receive.  So I actually this this method is fine, and there is no need to change it

Apart from ensuring the params are not None before checking if they are an existing file.

```
BotoServerError: 400 Bad Request\n<ErrorResponse xmlns=\"https://iam.amazonaws.com/doc/2010-05-08/\">
<Error>
    <Type>Sender</Type>
    <Code>MalformedCertificate</Code>
    <Message>Unable to parse certificate. Please ensure the certificate is in PEM format.</Message>
  </Error>
```

chrisdpa-tvx replied:
We can check if the string is a cert easily.  Certs are always in ASCII format (AWS doesn't accept binary certificates) and PEM certs will always start with `-----BEGIN`.  That would imply we can check for this as an identifier of a Cert (vs a file path)

chrisdpa-tvx replied:
@ProjectRice Sorry just saw your comment mentioning the same :)

ryansb replied:
I'd rather not try to implement cert validation, in case AWS changes what they accept.

s-hertel committed:
remove module used out of scope

check the potential filepath to be true before checking isfile

remove required: false

ProjectRice commented:
I think I'm happy for this to merge if after we fix the comment.  Open to further discussion about @ryansb's  
concern, however I don't think that's really an issue anymore given the failure message I got as above.

s-hertel commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -50,19 +50,19 @@
-      - Note that prior to 2.4, this parameter expected a path to a file.
-        Since 2.4 this is now accomplished using a lookup plugin. See examples for detail.
+      - The path to the CA certificate chain in PEM encoded format.
+        As of 2.4 this can also be accomplished using a lookup plugin.
```
Thanks! I'm modified it slightly because it's not clear which to use for versions prior to 2.4

s-hertel committed:
rephrase documentation

s-hertel committed:
remove 'lookups' from example

ProjectRice reviewed (approved):
I have now tested this in a variety of situations and I'm fairly happy we're good to go

Successful as expected
* local paths
* valid content
* mixture of local paths and content
* specifying cert_chain
* not specifying cert_chain

Failed as expected
* certs existing in IAM already
* cert names existing in IAM already
* Invalid cert content/path
* Invalid key content/path
* Invalid cert_chain content/path

AS discussed previously I can confirm AWS/Boto returns the following error in cases where users may specify a non existent path that is treated as content. As well as similar messages for keys and chains.

BotoServerError: 400 Bad Request
<Message>Unable to parse certificate. Please ensure the certificate is in PEM format.</Message>
